server/src/internalClusterTest/java/org/elasticsearch/action/support/master/TransportMasterNodeActionIT.java

```diff
@@ -107,24 +149,25 @@ public void onFailure(Exception e) {
                     }
                 });
 
-            safeAwait(electionWonLatch);
+            // Wait until the old master has acknowledged the new master's election
+            safeAwait(previousMasterKnowsNewMasterIsElectedLatch);
+            logger.info("New master is elected");
 
             // perform a TransportMasterNodeAction on the new master, which doesn't know it's the master yet
-            final var stateFuture = client(newMaster).admin().cluster().prepareState().clear().execute();
+            final var stateFuture = client(newMaster).admin().cluster().prepareHealth().execute();
 
-            // wait for the request to come back to the new master, which should now wait for its local term to advance
-            safeAwait(doubleReroutedMessageLatch);
+            // wait for the request to come back to the new master
+            safeAwait(newMasterReceivedReroutedMessageLatch);
+
+            // Unblock state application on new master, allow it to know of its election win
+            safeAwait(stateApplierBarrier);
 
             assertFalse(stateFuture.isDone());
 
-            releaseMasterLatch.countDown();
             safeGet(stateFuture);
-
         } finally {
-            enoughVotingMastersLatch.countDown();
-            electionWonLatch.countDown();
-            releaseMasterLatch.countDown();
-            internalCluster().getInstance(ClusterService.class, newMaster).removeApplier(blockingApplier);
+            // Unblock applier loop if it was left blocked
+            stateApplierBarrier.reset();
```

Comment:
`stateApplierBarrier` is the only synchroniser that a different thread waits on. All the others are waited on by the main thread which is clearly not blocked if we're here.